feat(config): implement custom header field inside HostRules (#26225)
renovatebot/renovate: docs/usage/self-hosted-configuration.md

## allowScripts

## allowedHeaders

> List of headers that are allowed to be forwarded to the HTTP request. This is useful when the `registryUrlTemplate` employs a specific authentication system. By default, all headers starting with "X-" are allowed, but you can include additional headers using this option. If declared, it will override the default "X-" allowed headers. `allowedHeaders` is an array of RegExp represented as strings due to JSON serialization

Reviewer: ```suggestion
List of headers that are allowed to be forwarded to the HTTP request. This is useful when the `registryUrlTemplate` employs a specific authentication system. By default, all headers starting with "X-" are allowed, but you can include additional headers using this option. If declared, it will override the default "X-" allowed headers. `allowedHeaders` is an array of minimatch-compatible globs or re2-compatible regex represented as strings.

Examples:
- `/X/` - any header with `abc` in the name (regex)
- `!/X/` - any header without `abcd` in the name (regex)
- `X-*` - any header starting with `X-` (glob pattern)
- `X` - only the header matching exactly `X` (exact match glob)
```

## allowScripts

## allowedHeaders

> List of headers that are allowed to be forwarded to the HTTP request. This is useful when the `registryUrlTemplate` employs a specific authentication system. By default, all headers starting with "X-" are allowed, but you can include additional headers using this option. If declared, it will override the default "X-" allowed headers. `allowedHeaders` is an array of minimatch-compatible globs or re2-compatible regex represented as strings.

Reviewer: ```suggestion
This is option particularly useful when a registry employs a specific authentication system not already covered by Renovate's standard credential handling in `hostRules`.
By default, all headers starting with "X-" are allowed, but you can permit additional headers using this option.
If declared, it will override the default "X-" allowed headers, so you should include them in your config if you wish for them to remain allowed.
`allowedHeaders` is an array of minimatch-compatible globs or re2-compatible regex strings.
```

Examples:

> - `/X/` - any header with `abc` in the name (regex)
> - `!/X/` - any header without `abcd` in the name (regex)

Reviewer: ```suggestion
- `/X/` - any header with `X` anywhere in the name (regex)
- `!/X/` - any header without `X` anywhere in the name (regex)
```

## allowScripts

## allowedHeaders

This is option particularly useful when a registry employs a specific authentication system not already covered by Renovate's standard credential handling in `hostRules`.
By default, all headers starting with "X-" are allowed, but you can permit additional headers using this option.
If declared, it will override the default "X-" allowed headers, so you should include them in your config if you wish for them to remain allowed.
`allowedHeaders` is an array of minimatch-compatible globs or re2-compatible regex strings.

Examples:

- `/X/` - any header with `X` anywhere in the name (regex)
- `!/X/` - any header without `X` anywhere in the name (regex)
- `X-*` - any header starting with `X-` (glob pattern)
- `X` - only the header matching exactly `X` (exact match glob)

```json
{
  "hostRules": [
    {
      "matchHost": "https://domain.com/all-versions",
      "headers": {
        "X-Auth-Token": "secret"
      }
    }
  ]
}
```

> or with custom `allowedHeaders`
> 
> ```json
> {
>   "allowedHeaders": ["custom-header"],
>   "hostRules": [
>     {
>       "matchHost": "https://domain.com/all-versions",
>       "headers": {
>         "custom-header": "secret"
>       }
>     }
>   ]
> }
> ```

Reviewer: ```suggestion
or with custom `allowedHeaders`:

```js title="config.js"
module.exports = {
  "allowedHeaders": ["custom-header"],
}
```

```json title="renovate.json"
{
  "hostRules": [
    {
      "matchHost": "https://domain.com/all-versions",
      "headers": {
        "custom-header": "secret"
      }
    }
  ]
}
```
```

## allowScripts

## allowedHeader

> This is option particularly useful when a registry employs a specific authentication system not already covered by Renovate's standard credential handling in `hostRules`.

Reviewer 2: ```suggestion
`allowedHeader` can be useful when a registry uses a authentication system that's not  covered by Renovate's default credential handling in `hostRules`.
```

> By default, all header starting with "X-" are allowed, but you can permit additional header using this option.

Reviewer 2: ```suggestion
By default, all headers starting with "X-" are allowed.
If needed, you can allow additional headers with the `allowedHeader` option.
```

Reviewer 3: it seems headers was right, should we use headers again?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers

Author: Should I change it everywhere? What about allowedHeaders, should I revert that as well?

> If declared, it will override the default "X-" allowed header, so you should include them in your config if you wish for them to remain allowed.

Reviewer 2: ```suggestion
Any set `allowedHeader` overrides the default "X-" allowed header, so you should include them in your config if you wish for them to remain allowed.
```

It's not clear to me what `included them` refers to. Overall this sentence is tough to understand.

> `allowedHeader` is an array of minimatch-compatible globs or re2-compatible regex strings.

Reviewer 2: ```suggestion
The `allowedHeader` config option takes an array of minimatch-compatible globs or re2-compatible regex strings.
```

Examples:

- `/X/` - any header with `X` anywhere in the name (regex)
- `!/X/` - any header without `X` anywhere in the name (regex)
- `X-*` - any header starting with `X-` (glob pattern)

> - `X` - only the header matching exactly `X` (exact match glob)

Reviewer 2: Please convert this to a table:

| Example header | Kind of pattern | Explanation |
| --- | --- | --- |
| `/X/` | Regex | Any header with `x` anywhere in the name |

And so on.

```json
{
  "hostRules": [
    {
      "matchHost": "https://domain.com/all-versions",
      "header": {
        "X-Auth-Token": "secret"
      }
    }
  ]
}
```

> or with custom `allowedHeader`:

Reviewer 2: ```suggestion
Or with custom `allowedHeader`:
```
